examples/vision/aodnet.py

```diff
@@ -0,0 +1,376 @@
+        )
+        original_image_files.append(
+            str(os.path.join(dataset_path, "original_images/" + original_file_name))
+        )
+    return original_image_files, hazy_image_paths
+
+
+def read_images(image_files):
+    dataset = tf.data.Dataset.from_tensor_slices(image_files)
+    dataset = dataset.map(tf.io.read_file)
+    dataset = dataset.map(
+        lambda x: tf.image.decode_png(x, channels=3),
+        num_parallel_calls=tf.data.experimental.AUTOTUNE,
+    )
+    return dataset
+
+
+def random_crop(hazy_image, original_image, low_crop_size, enhanced_crop_size):
+    hazy_image_shape = tf.shape(hazy_image)[:2]
+    low_w = tf.random.uniform(
+        shape=(), dtype=tf.int32, maxval=hazy_image_shape[1] - low_crop_size + 1
+    )
+    low_h = tf.random.uniform(
+        shape=(), dtype=tf.int32, maxval=hazy_image_shape[0] - low_crop_size + 1
+    )
+    enhanced_w = low_w
+    enhanced_h = low_h
+    hazy_image_cropped = hazy_image[
+        low_h : low_h + low_crop_size, low_w : low_w + low_crop_size
+    ]
+    original_image_cropped = original_image[
+        enhanced_h : enhanced_h + enhanced_crop_size,
+        enhanced_w : enhanced_w + enhanced_crop_size,
+    ]
+    return hazy_image_cropped, original_image_cropped
+
+
+def random_flip(hazy_image, original_image):
+    return tf.cond(
+        tf.random.uniform(shape=(), maxval=1) < 0.5,
+        lambda: (hazy_image, original_image),
+        lambda: (
+            tf.image.flip_left_right(hazy_image),
+            tf.image.flip_left_right(original_image),
+        ),
+    )
+
+
+def random_rotate(hazy_image, original_image):
+    condition = tf.random.uniform(shape=(), maxval=4, dtype=tf.int32)
+    hazy_image = tf.image.rot90(hazy_image, condition)
+    original_image = tf.image.rot90(original_image, condition)
+    return hazy_image, original_image
+
+
+def apply_scaling(hazy_image, original_image):
+    hazy_image = tf.cast(hazy_image, tf.float32)
+    original_image = tf.cast(original_image, tf.float32)
+    hazy_image = hazy_image / 255.0
+    original_image = original_image / 255.0
+    return hazy_image, original_image
+
+
+def configure_dataset(
+    dataset, image_crop_size, buffer_size, batch_size, is_dataset_train
+):
+    dataset = dataset.map(apply_scaling, num_parallel_calls=tf.data.AUTOTUNE)
+    if image_crop_size > 0:
+        dataset = dataset.map(
+            lambda hazy, original: random_crop(
+                hazy, original, image_crop_size, image_crop_size
+            ),
+            num_parallel_calls=tf.data.AUTOTUNE,
+        )
+    if is_dataset_train:
+        dataset = dataset.map(random_rotate, num_parallel_calls=tf.data.AUTOTUNE)
+        dataset = dataset.map(random_flip, num_parallel_calls=tf.data.AUTOTUNE)
+    dataset = dataset.shuffle(buffer_size)
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.repeat(1)
+    dataset = dataset.prefetch(buffer_size=tf.data.AUTOTUNE)
+    return dataset
+
+
+class DeHazeDataLoader:
```

Comment:
Something like should be a stateless function rather than a class.

The dataset generation code represents about 50% of the code example, but it could be considerably simplified via `image_dataset_from_directory`